### PR TITLE
Removing object mode fallback from `@jit`.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This removes the long deprecated "object mode fallback" behaviour found in the `numba.jit` decorator (and other similar decorators). The removal attempts to be "minimal" in case there are unknown issues with making this change with view of making it easy to revert.

### CODE REVIEW

File: numba/core/decorators.py [hunk 1]
```diff
@@ -143,34 +143,27 @@ def bar(x, y):
         raise DeprecationError(_msg_deprecated_signature_arg.format('argtypes'))
     if 'restype' in options:
         raise DeprecationError(_msg_deprecated_signature_arg.format('restype'))
-    if options.get('nopython', False) and forceobj:
+    nopython = options.get('nopython', None)
+    if nopython is True and forceobj:
```
Comment (sklam) [inline numba/core/decorators.py:147]: Does it matter that `nopython` is exactly the `True` object? or does it just need to be truthy?

File: numba/core/decorators.py [hunk 2]
```diff
@@ -143,34 +143,27 @@ def bar(x, y):
...
-                   "argument is currently False, but it will be changed to "
-                   f"True in Numba 0.59.0. See {url} for details.")
-        else:
-            msg = ("The keyword argument 'nopython=False' was supplied. From "
-                   "Numba 0.59.0 the default is being changed to True and use "
-                   "of 'nopython=False' will raise a warning as the "
-                   f"argument will have no effect. See {url} for "
-                   "details.")
-        warnings.warn(NumbaDeprecationWarning(msg), stacklevel=2)
+    if nopython is False:
```
Comment (sklam) [inline numba/core/decorators.py:155]: same question about truthiness here.

File: numba/core/decorators.py (hunk 1 above)
Comment (sklam) [reply numba/core/decorators.py:147]:
```python
In [4]: @jit(nopython=1, forceobj=True)
   ...: def foo(): return

In [5]: foo()
```

**Review by sklam — COMMENTED** [review]
Besides the CI failure on `test_globals_jit`, my only comment is below:

**Comment by stuartarchibald** [comment]
> Besides the CI failure on `test_globals_jit`, my only comment is below:

Hopefully fixed in 774dc75.

File: numba/core/decorators.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/decorators.py:149]: Assertion added to enforce `bool` in c3b8c33.

File: numba/core/decorators.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/decorators.py:157]: Assertion added to enforce `bool` in c3b8c33.

**Comment by stuartarchibald** [comment]
Note: the documentation/deprecation notices need updating to reflect this change.

**Comment by sklam** [comment]
The CI failure is showing more test failures. Many seems to be resolvable by adding `forceobj=True`. Some looplifting tests may no longer be valid since (esp the "reject" ones since they are testing fallback behavior).

In addition, we might need to remove `Flags.enable_pyobject`.

**Comment by gmarkall** [comment]
gpuci run tests

**Review by sklam — APPROVED** [review]

**Pull request merged**